source/readonly-deep.d.ts

```diff
@@ -0,0 +1,57 @@
+import {Primitive} from './basic';
+
+/**
+Convert `object`s, `Map`s, `Set`s, and `Array`s into immutable stuctures recursively. Converts all properties/contents into immutable structures recusively.
+
+This is useful when a deeply nested structure needs to be exposed as completely immutable, like for example an imported JSON module.
+
+@example
+```
+// data.json
+{
+	"foo": ["bar"]
+}
+
+// main.ts
+import {ReadonlyDeep} from 'type-fest';
+import dataJson = require('./data.json');
+
+const data: ReadonlyDeep<typeof dataJson> = dataJson;
+
+export default data;
+
+// test.ts
+import data from './main';
+
+data.foo.push('bar');
+//=> error TS2339: Property 'push' does not exist on type 'readonly string[]'
+```
+*/
+export type ReadonlyDeep<T> = T extends Primitive | ((...arguments: any[]) => unknown)
+	? T
+	: T extends ReadonlyMap<infer KeyType, infer ValueType>
+	? ReadonlyMapDeep<KeyType, ValueType>
+	: T extends ReadonlySet<infer ItemType>
+	? ReadonlySetDeep<ItemType>
+	: T extends object
+	? ReadonlyObjectDeep<T>
```

Comment:
Would there be any benefits in specifically targeting array here? Maybe improved debugging. Or would the result be exactly the same as now?

Reply:
Nope, TS 3.4 handles Arrays correctly with the `ReadonlyObjectDeep`. I've tried it out, it is actually harmful adding special casing for Arrays, it becomes very complicated to handle all variants of `readonly`, `Array`, and `[..., ...]` (tuples) correctly. The way it is right now works like a charm, all the special cases are green in the tests.